lib/_stream_writable.js

```diff
@@ -590,20 +585,18 @@ Writable.prototype.end = function(chunk, encoding, cb) {
   if (typeof cb !== 'function')
     cb = nop;
 
-  // Ignore unnecessary end() calls.
-  // TODO(ronag): Compat. Allow end() after destroy().
-  if (!state.errored && !state.ending) {
-    endWritable(this, state, cb);
-  } else if (state.finished) {
-    const err = new ERR_STREAM_ALREADY_FINISHED('end');
-    process.nextTick(cb, err);
-    // TODO(ronag): Compat. Don't error the stream.
-    // errorOrDestroy(this, err, true);
+  if (state.finished) {
+    throw new ERR_STREAM_ALREADY_FINISHED('end');
   } else if (state.destroyed) {
-    const err = new ERR_STREAM_DESTROYED('end');
-    process.nextTick(cb, err);
-    // TODO(ronag): Compat. Don't error the stream.
-    // errorOrDestroy(this, err, true);
+    // TODO(ronag): Compat. Throwing here can break stuff.
+    // throw new ERR_STREAM_DESTROYED('write');
+    // TODO(ronag): Compat. end() after destroy() must be allowed if
+    // done in same tick.
+    if (!state.ending) {
+      endWritable(this, state, cb);
+    }
```

Comment:
I don't remember what the root cause of this compat. Will need to dig into it again.